apps/api/src/purchase-orders/purchase-orders.service.ts

```diff
@@ -0,0 +1,202 @@
+        vendor_name: createPurchaseOrderDto.vendor_name,
+        order_date: createPurchaseOrderDto.order_date,
+        expected_delivery_date: createPurchaseOrderDto.expected_delivery_date,
+        purchase_order_line_items: {
+          create: createPurchaseOrderDto.purchase_order_line_items.map(item => ({
+            item_id: item.item_id,
+            quantity: item.quantity,
+            unit_cost: item.unit_cost,
+          })),
+        },
+      },
+    });
+  }
+
+  async update(id: number, updatePurchaseOrderDto: UpdatePurchaseOrderDto): Promise<PurchaseOrders> {    
+    // Extracting item IDs from the DTO
+    const updatedLineItemIds = updatePurchaseOrderDto.purchase_order_line_items.map(line_item => line_item.id);
+
+    // Step 1: Delete line items not present in the DTO
+    await this.prisma.purchaseOrderLineItems.deleteMany({
+      where: {
+        purchase_order_id: id,
+        id: { notIn: updatedLineItemIds },
+      },
+    });
+
+    // Step 2: Upsert line items
+    return this.prisma.purchaseOrders.update({
+      where: { id },
+      data: {
+        vendor_name: updatePurchaseOrderDto.vendor_name,
+        order_date: updatePurchaseOrderDto.order_date,
+        expected_delivery_date: updatePurchaseOrderDto.expected_delivery_date,
+        purchase_order_line_items: {
+          upsert: updatePurchaseOrderDto.purchase_order_line_items.map(line_item => ({
+            where: { id: line_item.id },
+            update: { item_id: line_item.item_id, quantity: line_item.quantity, unit_cost: line_item.unit_cost },
+            create: { item_id: line_item.item_id, quantity: line_item.quantity, unit_cost: line_item.unit_cost },
+          })),
+        },
+      },
+    });
+  }
+
+  async findOne(id: number): Promise<PurchaseOrders> {
+    const purchaseOrder = this.prisma.purchaseOrders.findUnique({
+      where: { id },
+      include: {
+        purchase_order_line_items: true, // Include related line items
+      },
+    });
+
+    if (!purchaseOrder) {
+      throw new NotFoundException(`Purchase Order with ID ${id} not found`);
+    }
+
+    return purchaseOrder;
+  }
+
+  /* 
+  A 3-way match occurs if the following are true:
+  1. Each item that is present in the purchase order is also present in the goods receipt and supplier invoice 
+  2. The total quantities of each item in the purchase order match the total quantities for that same item in the goods receipt and supplier invoice
+  3. The total cost of each item in the purchase order match the total cost for that same item in the supplier invoice  
+  */
+  async findMatches(): Promise<{ id: number, goodsReceiptMatch: boolean, supplierInvoiceMatch: boolean, reasons: string[] }[]> {
+    
+    // Step 1: Get relevant records from the database
+    const purchaseOrders = await this.prisma.purchaseOrders.findMany({
+      include: {
+        purchase_order_line_items: {
+          include: {
+            item: true
+          }
+        }
+      },
+    });
```

Comment:
Not important for this demo, but how would you control data access given we're building a multi-merchant app? I see your queries are essentially open, implying a single merchant. Just curious how you would approach this problem?

Reply:
First, you would need a way to identify which records belong to what merchant. One way to do this is to add a merchant_id to the relevant tables and the queries would need to be adjusted to use this merchant_id. 

The other thing you would need to solve for is how you handle what the merchant_id is for a particular user. This could be solved for by including the merchant_id in the authentication token. The merchant_id could be retrieved using middleware and used in the service layer.